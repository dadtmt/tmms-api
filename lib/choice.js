'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.newChoice = exports.isInteractive = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var isInteractive = exports.isInteractive = _ramda2.default.cond([[_ramda2.default.propEq('type', 'dice'), _ramda2.default.pipe(_ramda2.default.path(['content', 'master']), _ramda2.default.not)], [_ramda2.default.propEq('type', 'characterSheet'), _ramda2.default.F], [_ramda2.default.T, _ramda2.default.T]]);

var newChoice = exports.newChoice = function newChoice(values) {
  return _extends({}, values, {
    interactive: isInteractive(values),
    made: false
  });
};