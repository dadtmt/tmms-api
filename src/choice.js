import R from 'ramda'

import { getCurrentCrossroadStep } from './editor'

export const isInteractive = R.cond([
  [
    R.propEq('type', 'dice'),
    R.pipe(R.path(['content', 'master']), R.not)
  ],
  [R.propEq('type', 'characterSheet'), R.F],
  [R.T, R.T]
])

export const newChoice = values => data => ({
  ...values,
  interactive: isInteractive(values),
  made: false,
  step: getCurrentCrossroadStep(data)
})
