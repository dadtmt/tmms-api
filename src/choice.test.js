import { isInteractive, newChoice } from './choice'

describe('isInteractive', () => {
  it('return true by default', () => {
    const values = {}
    expect(isInteractive(values)).toBeTruthy()
  })
  it('return false if type is dice and master true', () => {
    const values = {
      content: {
        master: true
      },
      type: 'dice'
    }
    expect(isInteractive(values)).toBeFalsy()
  })
  it('return true if type is dice and master false', () => {
    const values = {
      content: {
        master: false
      },
      type: 'dice'
    }
    expect(isInteractive(values)).toBeTruthy()
  })
  it('return false if type is characterSheet', () => {
    const values = {
      type: 'characterSheet'
    }
    expect(isInteractive(values)).toBeFalsy()
  })
})

describe('newChoice', () => {
  const data = {
    viewer: {
      user: {
        editors: {
          edges: [{
            node: {
              crossroads: {
                edges: [{
                  node: {
                    id: 'CROSSROAD_ID'
                  }
                }]
              },
              id: 'EDITOR_ID'
            }
          }]
        }
      }
    }
  }
  it('copy values and add interactive and made choice', () => {
    const values = {
      some: 'values'
    }
    expect(newChoice(values)(data)).toMatchSnapshot()
  })
  it('copy values and add interactive false for characterSheet type', () => {
    const values = {
      some: 'values',
      type: 'characterSheet'
    }
    expect(newChoice(values)(data)).toMatchSnapshot()
  })
  it(
  'copy values and add interactive false for dice type and content.master true',
    () => {
      const values = {
        content: {
          master: true
        },
        some: 'values',
        type: 'dice'
      }
      expect(newChoice(values)(data)).toMatchSnapshot()
    }
  )
  it('set step to step 2 if step 1 is made', () => {
    const dataStep2 = {
      viewer: {
        user: {
          editors: {
            edges: [{
              node: {
                crossroads: {
                  edges: [{
                    node: {
                      choices: {
                        edges: [{
                          node: {
                            made: true,
                            step: 1
                          }
                        }]
                      },
                      id: 'CROSSROAD_ID'
                    }
                  }]
                },
                id: 'EDITOR_ID'
              }
            }]
          }
        }
      }
    }
    const values = {
      some: 'values'
    }
    expect(newChoice(values)(dataStep2)).toMatchSnapshot()
  })
})
