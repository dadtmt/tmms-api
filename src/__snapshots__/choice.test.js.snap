// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newChoice copy values and add interactive and made choice 1`] = `
Object {
  "interactive": true,
  "made": false,
  "some": "values",
  "step": 1,
}
`;

exports[`newChoice copy values and add interactive false for characterSheet type 1`] = `
Object {
  "interactive": false,
  "made": false,
  "some": "values",
  "step": 1,
  "type": "characterSheet",
}
`;

exports[`newChoice copy values and add interactive false for dice type and content.master true 1`] = `
Object {
  "content": Object {
    "master": true,
  },
  "interactive": false,
  "made": false,
  "some": "values",
  "step": 1,
  "type": "dice",
}
`;

exports[`newChoice set step to step 2 if step 1 is made 1`] = `
Object {
  "interactive": true,
  "made": false,
  "some": "values",
  "step": 2,
}
`;
